packages/contracts-bedrock/scripts/L2Genesis.s.sol

```diff
@@ -0,0 +1,373 @@
+                address implementation = _predeployToCodeNamespace(addr);
+                vm.store(addr, PROXY_IMPLEMENTATION_ADDRESS, bytes32(uint256(uint160(implementation))));
+            }
+        }
+    }
+
+    /// @notice LEGACY_ERC20_ETH is not being predeployed
+    function _setImplementations() internal {
+        _setLegacyMessagePasser();
+        _setDeployerWhitelist();
+        _setWETH9();
+        _setL2StandardBridge();
+        _setL2CrossDomainMessenger();
+        _setSequencerFeeVault();
+        _setOptimismMintableERC20Factory();
+        _setL1BlockNumber();
+        _setGasPriceOracle();
+        _setGovernanceToken();
+        _setL1Block();
+    }
+
+    function _setLegacyMessagePasser() internal {
+        address impl = _predeployToCodeNamespace(Predeploys.LEGACY_MESSAGE_PASSER);
+        vm.etch(
+            impl,
+            vm.getDeployedCode("LegacyMessagePasser.sol:LegacyMessagePasser")
+        );
+
+        _verifyProxyImplementationAddress(Predeploys.LEGACY_MESSAGE_PASSER, impl);
+    }
+
+    function _setDeployerWhitelist() internal {
+        address impl = _predeployToCodeNamespace(Predeploys.DEPLOYER_WHITELIST);
+        vm.etch(
+            _predeployToCodeNamespace(impl),
+            vm.getDeployedCode("DeployerWhitelist.sol:DeployerWhitelist")
+        );
+
+        _verifyProxyImplementationAddress(Predeploys.DEPLOYER_WHITELIST, impl);
+    }
+
+    /// @notice Contract metadata appended to deployed bytecode will differ
+    ///         from previous L2 genesis output. Additionally, this contract
+    ///         is NOT proxied.
+    function _setWETH9() internal {
+        vm.etch(
+            Predeploys.WETH9,
+            vm.getDeployedCode("WETH9.sol:WETH9")
+        );
+        vm.store(
+            Predeploys.WETH9,
+            // string public name
+            hex"0000000000000000000000000000000000000000000000000000000000000000",
+            // "Wrapped Ether"
+            hex"577261707065642045746865720000000000000000000000000000000000001a"
+        );
+        vm.store(
+            Predeploys.WETH9,
+            // string public symbol
+            hex"0000000000000000000000000000000000000000000000000000000000000001",
+            // "WETH"
+            hex"5745544800000000000000000000000000000000000000000000000000000008"
+        );
+        vm.store(
+            Predeploys.WETH9,
+            // uint8 public decimals
+            hex"0000000000000000000000000000000000000000000000000000000000000002",
+            // 18
+            hex"0000000000000000000000000000000000000000000000000000000000000012"
+        );
+    }
+
+    function _setL2StandardBridge() internal {
+        address impl = _predeployToCodeNamespace(Predeploys.L2_STANDARD_BRIDGE);
+        vm.etch(
+            impl,
+            vm.getDeployedCode("L2StandardBridge.sol:L2StandardBridge")
+        );
+        L2StandardBridge(payable(impl)).initialize(L1StandardBridge(payable(address(0))));
```

Comment:
For every initialize call, can we re-call it with an expectRevert? I think it should work as expected to do this directly in the script. The rationale is making sure that none of the implementations use a variable being zero (instead of a direct initializer slot) to allow `initialize` to be called

Reply:
Updated!

Reply:
Perhaps we have a `postCheck` function for all of this, altho this test case is covered because the output of this will be imported into the unit tests

Reply:
I can move these to a `postCheck` function. I like the idea of having these checks also in this script since it can be ran standalone without the unit tests being ran, just provides an extra safety check

Reply:
@tynes I didn't see a straightforward way to move the `vm.expectRevert` re-initialize checks since the `initialize` methods have different function args, so I opted to keep the checks in each `_set` function, but collected and moved the checks to the bottom of each function e.g.:

```solidity
function _setOptimismMintableERC20Factory() internal {
        address impl = _predeployToCodeNamespace(Predeploys.OPTIMISM_MINTABLE_ERC20_FACTORY);
        vm.etch(
            impl,
            vm.getDeployedCode("OptimismMintableERC20Factory.sol:OptimismMintableERC20Factory")
        );
        OptimismMintableERC20Factory(impl)
            .initialize(address(0));

        OptimismMintableERC20Factory(Predeploys.OPTIMISM_MINTABLE_ERC20_FACTORY)
            .initialize(Predeploys.L2_STANDARD_BRIDGE);

        /// Post Checks
        _verifyProxyImplementationAddress(Predeploys.OPTIMISM_MINTABLE_ERC20_FACTORY, impl);

        vm.expectRevert();
        OptimismMintableERC20Factory(impl)
            .initialize(address(0));

        vm.expectRevert();
        OptimismMintableERC20Factory(Predeploys.OPTIMISM_MINTABLE_ERC20_FACTORY)
            .initialize(Predeploys.L2_STANDARD_BRIDGE);
    }
```

Reply:
I added a `_verifyCantReInitialize` function to clean up the post checks a little, but didn't move everything into a single `postCheck`

```solidity
function _verifyCantReInitialize(address _contract, address _arg) internal {
        vm.expectRevert();
        IInitializable(_contract).initialize(_arg);
    }

function _setL2CrossDomainMessenger() internal {
        address impl = _predeployToCodeNamespace(Predeploys.L2_CROSS_DOMAIN_MESSENGER);
        vm.etch(
            impl,
            vm.getDeployedCode("L2CrossDomainMessenger.sol:L2CrossDomainMessenger")
        );
        L2CrossDomainMessenger(impl)
            .initialize(L1CrossDomainMessenger(address(0)));

        L2CrossDomainMessenger(Predeploys.L2_CROSS_DOMAIN_MESSENGER)
            .initialize(L1CrossDomainMessenger(mustGetAddress("L1CrossDomainMessengerProxy")));

        /// Post Checks
        _verifyProxyImplementationAddress(Predeploys.L2_CROSS_DOMAIN_MESSENGER, impl);
        _verifyCantReInitialize(impl, address(0));
        _verifyCantReInitialize(Predeploys.L2_CROSS_DOMAIN_MESSENGER, mustGetAddress("L1CrossDomainMessengerProxy"));
    }
```